probe-rs-cli-util/src/common_options.rs

```diff
@@ -0,0 +1,474 @@
+    )]
+    pub reset_halt: bool,
+    #[structopt(
+        name = "level",
+        long = "log",
+        help = "Use this flag to set the log level.\n\
+        Default is `warning`. Possible choices are [error, warning, info, debug, trace]."
+    )]
+    pub log: Option<log::Level>,
+    #[structopt(
+        name = "restore-unwritten",
+        long = "restore-unwritten",
+        help = "Enable this flag to restore all bytes erased in the sector erase but not overwritten by any page."
+    )]
+    pub restore_unwritten: bool,
+    #[structopt(
+        name = "filename",
+        long = "flash-layout",
+        help = "Requests the flash builder to output the layout into the given file in SVG format."
+    )]
+    pub flash_layout_output_path: Option<String>,
+    #[structopt(
+        name = "elf file",
+        long = "elf",
+        help = "The path to the ELF file to be flashed."
+    )]
+    pub elf: Option<String>,
+    #[structopt(
+        name = "directory",
+        long = "work-dir",
+        help = "The work directory from which cargo-flash should operate from."
+    )]
+    pub work_dir: Option<String>,
+    #[structopt(flatten)]
+    /// Arguments which are forwarded to 'cargo build'.
+    pub cargo_options: CargoOptions,
+    #[structopt(flatten)]
+    /// Argument relating to probe/chip selection/configuration.
+    pub probe_options: ProbeOptions,
+}
+
+impl FlashOptions {
+    pub fn early_exit(&self, f: impl Write) -> Result<bool, OperationError> {
+        if self.list_probes {
+            list_connected_probes(f)?;
+            return Ok(true);
+        }
+
+        if self.list_chips {
+            print_families(f)?;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+}
+
+#[derive(StructOpt, Debug)]
+pub struct ProbeOptions {
+    #[structopt(name = "chip", long = "chip")]
+    pub chip: Option<String>,
+    #[structopt(name = "chip description file path", long = "chip-description-path")]
+    pub chip_description_path: Option<String>,
+    #[structopt(name = "protocol", long = "protocol", default_value = "swd")]
+    pub protocol: WireProtocol,
+    #[structopt(
+        long = "probe",
+        help = "Use this flag to select a specific probe in the list.\n\
+        Use '--probe VID:PID' or '--probe VID:PID:Serial' if you have more than one probe with the same VID:PID."
+    )]
+    pub probe_selector: Option<DebugProbeSelector>,
+    #[structopt(
+        long = "connect-under-reset",
+        help = "Use this flag to assert the nreset & ntrst pins during attaching the probe to the chip."
+    )]
+    pub connect_under_reset: bool,
+    #[structopt(name = "speed", long = "speed", help = "The protocol speed in kHz.")]
+    pub speed: Option<u32>,
+    #[structopt(long = "dry-run")]
+    pub dry_run: bool,
+}
+
+static mut PROBE: Option<Probe> = None;
+static mut TARGET_SELECTOR: Option<TargetSelector> = None;
```

Comment:
Does the trick, but is there a better way? Additionally, the module now contains state, and functions are expected to be called in a certain order.

Reply:
Expected order is
1. `FlashOptions::maybe_load_chip_desc -> Result<()>`.
2. `FlashOptions::early_exit -> Result<bool>`: whether application should exit.
3. `FlashOptions::get_target_selector -> Result<&TargetSelector>`: `TARGET_SELECTOR` set.
4. `FlashOptions::attach_to_probe -> Result<&Probe>`: `PROBE` set.
5. `FlashOptions::attach -> Result<Session>`: `PROBE` consumed and cleared.

Steps 1 should probably be merged into 2. Steps 3--5 can be merged together so as to hide any unsafe access from the user, but then we need to add functions to query metadata. Thoughts?

Reply:
I think this can be made safe with an `Arc`.

Reply:
For me, `ProbeOptions::attach` does not a lot of work, so I would prefer to just remove it. This creates a bit more work for the consumer of this crate, but it makes the implementation a lot easier. We could also add an additional convenience method, which performs `ProbeOptions::attach_to_probe` and `ProbeOptions::attach` together.

If we want to keep it, I think we could move the fields onto the `ProbeOptions` struct itself. This avoids the issue with global state, and allows us to keep the same API.